C++ APIs CUDA Stream Note

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45754 C++ APIs CUDA Stream Note**

This is a doc to teach user how to use CUDA Stream in general. Will add more in the future for advanced topics.

![Screen Shot 2020-10-06 at 5 25 34 PM](https://user-images.githubusercontent.com/32228059/95261967-550b7f00-07f9-11eb-9a02-2cce62682db1.png)
![Screen Shot 2020-10-06 at 5 25 56 PM](https://user-images.githubusercontent.com/32228059/95261971-576dd900-07f9-11eb-9c6e-9d2dd206012c.png)
![Screen Shot 2020-10-06 at 5 26 07 PM](https://user-images.githubusercontent.com/32228059/95261975-59379c80-07f9-11eb-95ca-d79edeca470c.png)
![Screen Shot 2020-10-06 at 5 26 19 PM](https://user-images.githubusercontent.com/32228059/95261978-5a68c980-07f9-11eb-8f8a-71fb8cb8c0fa.png)
![Screen Shot 2020-10-06 at 5 26 29 PM](https://user-images.githubusercontent.com/32228059/95261981-5c328d00-07f9-11eb-988f-07ff6424994e.png)




Differential Revision: [D24085103](https://our.internmc.facebook.com/intern/diff/D24085103)

Commit:
C++ APIs CUDA Stream Note (Set/Get part)

[ghstack-poisoned]

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,189 @@
+Tensor CUDA Stream API 
+======================
+
+A `CUDA Stream`_ is a linear sequence of execution that belongs to a specific device.
```
"specific device" -> "specific CUDA device"? Just in case.

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,189 @@
+======================
+
+A `CUDA Stream`_ is a linear sequence of execution that belongs to a specific device.
+Pytorch C++ API has defined a CUDAStream class and a few useful functions to support CUDA Stream operations.
```
What if this sentence was a little more positive?

"The PyTorch C++ API supports CUDA streams with the CUDAStream class and useful helper functions to make streaming operations easy."

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,189 @@
+.. _CUDAStream.h: https://pytorch.org/cppdocs/api/file_c10_cuda_CUDAStream.h.html#file-c10-cuda-cudastream-h
+.. _CUDAStreamGuard.h: https://pytorch.org/cppdocs/api/structc10_1_1cuda_1_1_c_u_d_a_stream_guard.html
+
+Get CUDA stream
```
Maybe "Acquiring CUDA Streams"?

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,189 @@
+Get CUDA stream
+*****************
+
+Pytorch C++ API provides the following ways to get CUDA stream:
```
If the above change is made, then the end of this sentence would change to:

"... to acquire a CUDA stream:"

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,189 @@
+
+.. code-block:: cpp
+  
+  CUDAStream getStreamFromPool(const bool isHighPriority = false, DeviceIndex device = -1);
```
This is not an issue for this PR, but in general it's too bad that we're inconsistent about the naming of our DeviceIndex parameters. That is, I'm sure users would expect us to use device_index here like we do in getDefaultCUDAStream.

glaringlee replied:
Yes, device and device_index are used inconsistently among cuda stream/guard/util apis, not nice, need a fix.

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,189 @@
+.. tip::
+
+  You can request a stream from the high priority pool by setting isHighPriority to true, or a stream for a specific device
+  by setting device (defaulting to the current CUDA stream).
```
maybe "by passing a device index" to be more consistent with the language below? 

By default I think this returns a low priority stream on the current device, but not necessarily the current CUDA stream. Or am I misremember how this works?

glaringlee replied:
This should be "defauting to the current CUDA stream's device index)

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,189 @@
+  
+  The default stream is where most computation occurs when you aren't explicitly using streams.
+
+3. Get the current CUDA stream, for the passed CUDA device, or for the current device if no device index is passed. 
```
"for the passed CUDA device" -> "for the CUDA device with index device_index, ..."

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,189 @@
+
+Pytorch C++ API provides the following ways to set CUDA stream:
+
+1. Set the current stream on the device of the passed in stream to be the passed in stream.
```
I have some regrets about designing the API like this because the scenario of setting a CUDA stream but not switching to it is rare and this has been confusing for people. Maybe in the future we should offer switchToCUDAStream(CUDAStream stream) which sets the device to the stream's device and sets the stream on that device to the given stream?

glaringlee replied:
Yes, this is strange. I think stream guard is used to switch both device and stream. Adding switchToCUDAStream is probably more strightforward.

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,189 @@
+
+.. tip::
+
+  This function may have **nothing** to do with the current device, it toggles the current stream of the device of the passed stream.
```
This and the next note are helpful.

mruberry replied:
I think the notes may be combined like this:

"This function may have **nosthing** to do with the current device. It only changes the current stream on the stream's device. We recommend using CUDAStreamGuard, instead, since it switches to the stream's device and makes it the current stream on that device.  CUDAStreamGuard will also restore the current device and stream when it's destroyed."

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,189 @@
+  We prefer people to use ``CUDAStreamGuard`` (which will switch both your current device and current stream in the way you expect, and
+  reset it back to its original state afterwards) instead of ``setCurrentCUDAStream``.
+
+2. Use ``CUDAStreamGuard`` to set CUDA stream within a scope, it is defined in `CUDAStreamGuard.h`_
```
"... to switch to a CUDA stream within a scope, ..." to use language like the above note.

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,189 @@
+2. Use ``CUDAStreamGuard`` to set CUDA stream within a scope, it is defined in `CUDAStreamGuard.h`_
+
+.. tip::
+   If you need to set streams on multiple devices on CUDA, use ``CUDAMultiStreamGuard`` instead.
```
"Use CUDAMultiStreamGuard if you need to streams on multiple CUDA devices."

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,189 @@
+CUDA Stream Usage Examples
+**************************
+
+1. Get and Set CUDA stream
```
"Acquiring and setting a CUDA stream" might sound more natural

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,189 @@
+
+.. code-block:: cpp
+
+  // get a new stream from CUDA stream pool on current device (assume current device is device 0)
```
Why is the assumption that the current device is device 0 needed?

glaringlee replied:
I just want to make this clear, and assuming there is no previous code setting default device to 1 for eg.

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,189 @@
+  // get a new stream from CUDA stream pool on current device (assume current device is device 0)
+  torch::cuda::CUDAStream myStream = at::cuda::getStreamFromPool();
+  // set myStream as the current stream on myStream's current device (device 0 as well)
+  // note that myStream's current device is the same as current device (within scope) here
```
I think a single brief comment at the start of this block might be more readable. The code is already very readable.

glaringlee replied:
I will re-org the comments in the examples.

Commit:
Update on "C++ APIs CUDA Stream Note (Get and Set Part)"


This is a doc to teach user how to use CUDA Stream in general. Will add more in the future for advanced topics.

![Screen Shot 2020-10-02 at 2 07 39 PM](https://user-images.githubusercontent.com/32228059/94955615-db058e00-04b8-11eb-9601-4f65bde1ac85.png)
![Screen Shot 2020-10-02 at 2 08 36 PM](https://user-images.githubusercontent.com/32228059/94955626-de007e80-04b8-11eb-94da-fb62de7a05e8.png)
![Screen Shot 2020-10-02 at 2 08 54 PM](https://user-images.githubusercontent.com/32228059/94955636-e062d880-04b8-11eb-811c-843339014d8e.png)


Differential Revision: [D24085103](https://our.internmc.facebook.com/intern/diff/D24085103)

[ghstack-poisoned]

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,173 @@
+
+.. attention::
+  
+  Above code is running on the same device/gpu (say device 0). `setCurrentCUDAStream` works as expected and set current CUDA stream correctly
```
I think this is a little misleading because the above code snippets will work no matter what device you're on.

glaringlee replied:
Ah, what I want to say is that this is setCurrentCUDAStream will have strange behavior if passed in stream's device is not the same as current device. Will change a bit.

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,173 @@
+
+.. code-block:: cpp
+
+   // create a vector of CUDA stream from current device (device 0)
```
I'm not really sure I get what this example is supposed to show?

glaringlee replied:
This example is to list different usage of guard.  It listed the changes of streams and devices in each step to demonstrate how the guard works.

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,173 @@
+     streams1.push_back(torch::cuda::getDefaultCUDAStream());
+     // get a new stream from CUDA stream pool on device 1
+     streams1.push_back(torch::cuda::getStreamFromPool());
+   }
```
You may want to add a comment explaining what the device guard is doing here. That is, it switches to device 1, then when it goes out of scope it resets the current device to device 0.

glaringlee replied:
I think previous comments are enough, but I will add these.

> This is a helpful document for people trying to use streams with the C++ API. The one part of it that may need a change is the examples. Maybe something like:
> 
> * a combination of your first two examples, where you acquire a stream, switch to it with setCurrentCUDAStream, say that you'd do some work here, and then switch back to the default stream. Also, do you want to show events here, too?
> * an example where you use a stream guard to do the same thing
> * an example where you use a stream guard. to also change the device
> * an example where you use a multistream guard
> 
> One narrative to motivate the multi-device setting may just be "now suppose you want to do work on the non-default stream on two devices simultaneously, and we already have streams on both devices in a vector, then..."
> 
> What are your thoughts?

Yes, I can do it. 
I think there is a misunderstanding here :).  I was not aiming to make this as a tutorial, but a centralized tech doc for programming people to use CUDA stream. So i focused on what action will result in what behavior, user should be able to mixed these actions to meet their needs.

I get your point, I'll re-org the examples and add more examples as you suggested, I think that make sense.

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,173 @@
+   assert(torch::cuda::current_device() == 0);
+   assert(torch::cuda::getCurrentCUDAStream(1) == streams1[0]);
+
+   // CUDA Device guard on changes the current device, but not the stream
```
This is a little misleading. Changing the device does change the current stream, after all (to whatever stream is current on the device switched to).

glaringlee replied:
typo.....I want to say 'CUDA Device guard only changes the current device'

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,173 @@
+   // current device is still 0 out of the scope of previous device guard
+   assert(torch::cuda::current_device() == 0);
+
+   // use CUDAStreamGuard to set a stream, it changes the stream on curernt devide and also passed in stream's device
```
"on curernt devide" -> "on the current device"

mruberry reviewed (commented):
This is a helpful document for people trying to use streams with the C++ API. The one part of it that may need a change is the examples. Maybe something like:

- a combination of your first two examples, where you acquire a stream, switch to it with setCurrentCUDAStream, say that you'd do some work here, and then switch back to the default stream. Also, do you want to show events here, too?
- an example where you use a stream guard to do the same thing
- an example where you use a stream guard. to also change the device
- an example where you use a multistream guard

One narrative to motivate the multi-device setting may just be "now suppose you want to do work on the non-default stream on two devices simultaneously, and we already have streams on both devices in a vector, then..."

What are your thoughts?

Commit:
Update on "C++ APIs CUDA Stream Note (Get and Set Part)"


This is a doc to teach user how to use CUDA Stream in general. Will add more in the future for advanced topics.

![Screen Shot 2020-10-02 at 2 07 39 PM](https://user-images.githubusercontent.com/32228059/94955615-db058e00-04b8-11eb-9601-4f65bde1ac85.png)
![Screen Shot 2020-10-02 at 2 08 36 PM](https://user-images.githubusercontent.com/32228059/94955626-de007e80-04b8-11eb-94da-fb62de7a05e8.png)
![Screen Shot 2020-10-02 at 2 08 54 PM](https://user-images.githubusercontent.com/32228059/94955636-e062d880-04b8-11eb-811c-843339014d8e.png)


Differential Revision: [D24085103](https://our.internmc.facebook.com/intern/diff/D24085103)

[ghstack-poisoned]

glaringlee commented:
> This is a helpful document for people trying to use streams with the C++ API. The one part of it that may need a change is the examples. Maybe something like:
> 
> * a combination of your first two examples, where you acquire a stream, switch to it with setCurrentCUDAStream, say that you'd do some work here, and then switch back to the default stream. Also, do you want to show events here, too?
> * an example where you use a stream guard to do the same thing
> * an example where you use a stream guard. to also change the device
> * an example where you use a multistream guard
> 
> One narrative to motivate the multi-device setting may just be "now suppose you want to do work on the non-default stream on two devices simultaneously, and we already have streams on both devices in a vector, then..."
> 
> What are your thoughts?

Yes, I can do it.
I think there is a misunderstanding here :). I was not aiming to make this as a tutorial, but a centralized tech doc for programming people to use CUDA stream. So i focused on what action will result in what behavior, user should be able to mixed these actions to meet their needs.

I get your point, I'll re-org the examples and add more examples as you suggested, I think that make sense.

Commit:
Update on "C++ APIs CUDA Stream Note (Get and Set Part)"


This is a doc to teach user how to use CUDA Stream in general. Will add more in the future for advanced topics.

![Screen Shot 2020-10-02 at 2 07 39 PM](https://user-images.githubusercontent.com/32228059/94955615-db058e00-04b8-11eb-9601-4f65bde1ac85.png)
![Screen Shot 2020-10-02 at 2 08 36 PM](https://user-images.githubusercontent.com/32228059/94955626-de007e80-04b8-11eb-94da-fb62de7a05e8.png)
![Screen Shot 2020-10-02 at 2 08 54 PM](https://user-images.githubusercontent.com/32228059/94955636-e062d880-04b8-11eb-811c-843339014d8e.png)


Differential Revision: [D24085103](https://our.internmc.facebook.com/intern/diff/D24085103)

[ghstack-poisoned]

glaringlee commented:
@mruberry 
I did some re-orgs to the example part.
The first 3 examples are specifically describe how to use CUDA stream C++ APIs in different aspects.
The last example is a skeleton(framework) example which I  do think developer can get more benefit from.
pls take a look at, thank you!

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,277 @@
+Acquiring CUDA stream
+*********************
+
+Pytorch C++ API provides the following ways to acquire CUDA stream:
```
"Pytorch" -> "PyTorch's"

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,277 @@
+  You can request a stream from the high priority pool by setting isHighPriority to true, or a stream for a specific device
+  by setting device index (defaulting to the current CUDA stream's device index).
+
+2. Acquire the default CUDA stream, for the passed CUDA device, or for the current device if no device index is passed.
```
"stream, for" -> "stream for" (the first comma isn't needed)

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,277 @@
+Set CUDA stream
+***************
+
+Pytorch C++ API provides the following ways to set CUDA stream:
```
"Pytorch" -> "PyTorch's"

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,277 @@
+
+.. code-block:: cpp
+
+  // This example shows how to acquire and set CUDA stream on the same device.
```
Nice example.

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,277 @@
+
+.. code-block:: cpp
+
+  // This example is the same as previous example, but explicitly specify device
```
This example makes a lot of sense.

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,277 @@
+
+.. attention::
+  
+  Above code is running on the same CUDA device. `setCurrentCUDAStream` will always set current CUDA stream on current device, but please
```
"but please pay attention on that" -> "but note that"

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,277 @@
+  pay attention on that `setCurrentCUDASteram` actually set current stream on the device of passed in CUDA stream.
+  
+
+2. Acquiring and setting CUDA stream on multi-devices.
```
"stream on multi-devices" -> "streams on multiple devices"

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,277 @@
+  // sum() on tensor0 use `myStream0` as current CUDA stream on device 0
+  tensor0.sum();
+
+  // change the current device index to 1 by using CUDA device guard within a braket scope
```
"braket" -> "curly brace" OR you can just remove the word and say "... within a scope"

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,277 @@
+  each passed in stream's device. Other than scope controlling, this guard is equivalent to 
+  calling ``setCurrentCUDAStream`` on each passed in stream.
+
+4. A skeleton example for handling CUDA streams on multi-devices
```
"multi-devices" -> "multiple devices"

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,277 @@
+
+.. code-block:: cpp
+
+   // This is a skeleton example that shows how to handle CUDA streams on multi-devices
```
"multi-devices" -> "multiple devices"

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,277 @@
+
+   // This is a skeleton example that shows how to handle CUDA streams on multi-devices
+   // Suppose you want to do work on the non-default stream on two devices simultaneously, and we
+   // already have streams on both devices in two vectors. The following code shows possible ways
```
"shows possible ways of acquiring, setting CUDA streams." -> "shows three ways of acquiring and setting the streams."

mruberry commented on docs/cpp/source/notes/tensor_cuda_stream.rst:
```diff
@@ -0,0 +1,277 @@
+   // This is a skeleton example that shows how to handle CUDA streams on multi-devices
+   // Suppose you want to do work on the non-default stream on two devices simultaneously, and we
+   // already have streams on both devices in two vectors. The following code shows possible ways
+   // of acquiring, setting CUDA streams. You can use this example as a reference and find useful
```
I would eliminate the last sentence starting with "You can use..."

mruberry reviewed (approved):
This is a cool note that I expect will help developers looking to better understand and work with streams using PyTorch's C++ API. I made some phrasing suggestions but overall looks good!

Commit:
Update on "C++ APIs CUDA Stream Note"


This is a doc to teach user how to use CUDA Stream in general. Will add more in the future for advanced topics.

![Screen Shot 2020-10-05 at 10 43 59 AM](https://user-images.githubusercontent.com/32228059/95094605-13df7600-06f8-11eb-8b11-13a18fe790c6.png)
![Screen Shot 2020-10-05 at 10 44 30 AM](https://user-images.githubusercontent.com/32228059/95094660-235ebf00-06f8-11eb-91d5-21a54675d3ea.png)
![Screen Shot 2020-10-05 at 10 44 42 AM](https://user-images.githubusercontent.com/32228059/95094669-2659af80-06f8-11eb-99cd-f30ee193a795.png)
![Screen Shot 2020-10-05 at 10 44 51 AM](https://user-images.githubusercontent.com/32228059/95094679-278adc80-06f8-11eb-91aa-432a7f71b277.png)
![Screen Shot 2020-10-05 at 10 45 02 AM](https://user-images.githubusercontent.com/32228059/95094697-2d80bd80-06f8-11eb-8d7d-6e0724b30cec.png)



Differential Revision: [D24085103](https://our.internmc.facebook.com/intern/diff/D24085103)

[ghstack-poisoned]

litttl commented:
I follow the APIS and use CUDAStream in thread function ,however it doesn't work well. Could you give me some guide?

glaringlee commented:
@litttl 
Yes, we can help. Please create a post in the forum c++ category here:
https://discuss.pytorch.org/c/c/11, we tend to answer these questions in the forum.
 
Please provide detailed info, so we can help. One thing you need to pay attention on the high level is that you need to sync cuda events if working with multi CUDA streaming to make sure the run order is what you want. And the Guard is per thread, not cross thread.
